### ISSUE TITLE
Parallel computation of SINR or received power for different steering angles

### ISSUE THREAD

**Issue by yuyuyuyao-ai** [NVlabs/sionna#1133]
In my current implementation, I run RadioMapSolver separately for each steering angle. It works but requires substantial computational overhead. Are there supported approaches to parallelize or batch these evaluations to reduce runtime?

Specifically, can RadioMapSolver generate SINR maps for multiple co-located transmitters that share identical array geometry, position, and orientation, but differ only in their precoding_vec (i.e., beamforming weights)? This would enable evaluating multiple beam directions within a single solver run. In my use case, I require custom beamforming weights, so modifying transmitter orientation to emulate beam steering is not a viable option.

If this is not supported, would an alternative workflow—such as running PathSolver once to obtain propagation paths and then applying beamforming weights offline—be more efficient? I am flexible with the performance metric, as long as I can evaluate received power for different steering angles.

Thanks!

[sionnar_rt_beamforming_debug.ipynb](https://github.com/user-attachments/files/26149105/sionnar_rt_beamforming_debug.ipynb)

**Comment by faycalaa** [comment]
Hi @yuyuyuyao-ai ,

You can specify a different precoding vector for each transmitter by providing a matrix of shape `[num_tx, num_tx_ant]`, where each row contains the precoding vector for one transmitter.

**Comment by yuyuyuyao-ai** [comment]
I tested this method and got an issue. With a single transmitter, the computed SINR values are as expected. However, for two transmitters with each using a different steering angle, the resulting SINR values deviate significantly from the single-transmitter case.

This is the result for one transmitter
"Steering Angles: [0]
num_tx = 1
SINR metrics are [**74.9838**]"


This is the result for two transmitters
"Steering Angles: [0, 15]
num_tx = 2
SINR metrics are [**-3.2739685**  3.2739677]"


[sionnar_rt_beamforming_debug.ipynb](https://github.com/user-attachments/files/26250691/sionnar_rt_beamforming_debug.ipynb)

**Comment by llyaljl** [comment]
I have a new question: Sionna RT is built on Mitsuba 3 and Dr.Jit. These frameworks provide automatic differentiation and can backpropagate gradients throughout the entire system. I want to use the built-in automatic differentiation of Sionna RT. How should I use it? Can you provide a reference?

> 你好，
> 
> 你可以通过提供形状矩阵，为每个发射器指定不同的预编码向量，其中每一行包含一个发射器的预编码向量。`[num_tx, num_tx_ant]`

**Comment by yuyuyuyao-ai** [comment]
> I have a new question: Sionna RT is built on Mitsuba 3 and Dr.Jit. These frameworks provide automatic differentiation and can backpropagate gradients throughout the entire system. I want to use the built-in automatic differentiation of Sionna RT. How should I use it? Can you provide a reference?
> 
> > 你好，
> > 你可以通过提供形状矩阵，为每个发射器指定不同的预编码向量，其中每一行包含一个发射器的预编码向量。`[num_tx, num_tx_ant]`

Hi @llyaljl , It looks like your question was posted in the wrong thread. This discussion is not related to gradients. You may find this link helpful for guidance on using Sionna’s built-in gradient functionality, https://nvlabs.github.io/sionna/rt/developer/dev_compat_frameworks.html#gradients

**Comment by yuyuyuyao-ai** [comment]
Hi @faycalaa, using this method, I would need to define num_tx co-located transmitters if I want to evaluate num_tx steering angles simultaneously. However, if those transmitters radiate at the same time, the resulting SINR would differ from the single-transmitter case for multiple steering angles. Is that correct?

> Hi [@yuyuyuyao-ai](https://github.com/yuyuyuyao-ai) ,
> 
> You can specify a different precoding vector for each transmitter by providing a matrix of shape `[num_tx, num_tx_ant]`, where each row contains the precoding vector for one transmitter.

**Comment by yuyuyuyao-ai** [comment]
Any update?

**Comment by alonekite** [comment]
Hi,

For evaluating multiple beamforming directions at specific receiver locations, you do not need to run the ray tracer separately for each beam.

A practical workflow is:

1. Run `PathSolver` once for the scene and receiver location.
2. Compute the channel frequency response from the resulting `paths`.
3. Apply all beamforming vectors offline using a batched matrix multiplication.
4. Calculate the received power for each beam and, if needed, average across subcarriers.

For example:

```python
import mitsuba as mi
import numpy as np
import torch

# Run ray tracing once.
paths = path_solver(scene, ...)

# Compute the frequency response for all desired subcarriers.
h_freq = paths.cfr(
    frequencies=mi.Float(frequencies_hz),
    normalize=False,
    normalize_delays=False,
    out_type="numpy",
)

# For one TX, one RX, and one time step:
# raw shape: [1, num_rx_ports, 1, num_tx_ports, 1, num_subcarriers]
raw = np.asarray(h_freq)

# H shape: [num_subcarriers, num_rx_ports, num_tx_ports]
H = torch.from_numpy(
    raw[0, :, 0, :, 0, :]
).permute(2, 0, 1).to(torch.complex64)

# W shape: [num_beams, num_tx_ports]
# Each row contains one precoding vector in Sionna's TX port ordering.
W = beam_weights.to(dtype=torch.complex64, device=H.device)

# Effective channel for every subcarrier and beam:
# [num_subcarriers, num_beams, num_rx_ports]
h_eff = torch.einsum("krt,bt->kbr", H, W)

# Received channel power with RX-port combining:
# [num_subcarriers, num_beams]
beam_power = h_eff.abs().square().sum(dim=-1)

# Average received channel power across subcarriers:
# [num_beams]
mean_beam_power = beam_power.mean(dim=0)

best_beam_index = mean_beam_power.argmax().item()
```

I used this approach with a 128-port dual-polarized transmitter, a two-port receiver, 3,276 active subcarriers, and 1,024 candidate beams. All beams were evaluated from the same ray-tracing result.

One important distinction: this works for evaluating beams at specific receiver positions. It does not automatically produce a complete spatial radio map for every beam. If your goal is a separate full coverage map for each precoding vector, `RadioMapSolver` may still need to be executed separately for each beam unless it provides a supported batched-precoding interface.

Also, make sure the beamforming vectors follow Sionna’s physical antenna-port ordering before applying them to the CFR.

A complete implementation is available here: https://github.com/alonekite/mMIMO_sleep
